### PR TITLE
Updated hook and middleware encapsulation

### PR DESCRIPTION
**Pull request by delvedor**
With this change we are able to register hooks and middlewares declared in a plugin *after* a route declaration in the same level.
The context update is postponed until `avvio` emits the `preReady` event (https://github.com/mcollina/avvio/pull/49).
I've also updated the hook runner since the context binding happens during the hook registration.

When using the hooks the benchmarks have improved by 1-2k req/sec.

Related: https://github.com/fastify/fastify/issues/748

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: fastify.js [hunk 1]
```diff
@@ -80,7 +80,7 @@ function build (options) {
   var listening = false
   // true when Fastify is ready to go
   var started = false
-  app.on('start', () => {
+  app.on('preReady', () => {
```
Comment (mcollina) [inline fastify.js]: can you please revert this to `'start'`? Why did you change it?

Comment (delvedor) [reply fastify.js]: Ooops, a little forgetfulness :P

File: fastify.js [hunk 2]
```diff
@@ -471,7 +473,7 @@ function build (options) {
         _fastify._contentTypeParser,
         config,
         _fastify._errorHandler,
-        buildMiddie(_fastify._middlewares),
+        null,
```
Comment (jsumners) [inline fastify.js]: With this PR, is this parameter still necessary?

Comment (delvedor) [reply fastify.js]: Yes, here:
https://github.com/fastify/fastify/blob/c0c58d68c37aebe7e2b2306451a6a99e5d927fb1/fastify.js#L655-L668

File: fastify.js [hunk 3]
```diff
@@ -501,6 +503,23 @@ function build (options) {
         return
       }
 
+      // It can happen that a user register a plugin with some hooks/middlewares *after*
+      // the route registration. To be sure to load also that hoooks/middlwares,
+      // we must listen for the avvio's preReady event, and update the context object accordingly.
+      app.once('preReady', () => {
```
Comment (nwoltman) [inline fastify.js]: Why not use this on line [442](https://github.com/fastify/fastify/pull/762/files#diff-20da0b6140a4768d182af0d9b185ac28R442) to replace  `_fastify.after(...`?

Comment (delvedor) [reply fastify.js]: Because if we do it we will throw schema and route errors outside the `avvio`'s lifecycle.

Comment (nwoltman) [reply fastify.js]: Oh that makes sense :)
